cmd/disk-cache.go

```diff
@@ -0,0 +1,674 @@
+	// Used for Content-Length
+	Size int64 `json:"size"`
+	// True if the object was cached on an anonymous request
+	// Used when the backend is down and we have to know if
+	// we can serve the current anonymous request
+	Anonymous bool `json:"anonymous"`
+	// Other user defined metadata
+	HTTPMeta map[string]string `json:"httpMeta"`
+}
+
+// Converts diskCacheObjectMeta to ObjectInfo.
+func (objectMeta diskCacheObjectMeta) toObjectInfo() ObjectInfo {
+	objInfo := ObjectInfo{}
+	objInfo.UserDefined = make(map[string]string)
+	objInfo.Bucket = objectMeta.BucketName
+	objInfo.Name = objectMeta.ObjectName
+
+	objInfo.ModTime = objectMeta.ModTime
+	objInfo.Size = objectMeta.Size
+	objInfo.ETag = objectMeta.HTTPMeta["etag"]
+	objInfo.ContentType = objectMeta.HTTPMeta["content-type"]
+	objInfo.ContentEncoding = objectMeta.HTTPMeta["content-encoding"]
+
+	for key, val := range objectMeta.HTTPMeta {
+		if key == "etag" {
+			continue
+		}
+		objInfo.UserDefined[key] = val
+	}
+
+	return objInfo
+}
+
+// Converts objInfo to diskCacheObjectMeta.
+func newDiskCacheObjectMeta(objInfo ObjectInfo, anon bool) diskCacheObjectMeta {
+	objMeta := diskCacheObjectMeta{}
+	objMeta.HTTPMeta = make(map[string]string)
+	objMeta.BucketName = objInfo.Bucket
+	objMeta.ObjectName = objInfo.Name
+	objMeta.ModTime = objInfo.ModTime
+	objMeta.Size = objInfo.Size
+
+	objMeta.HTTPMeta["etag"] = objInfo.ETag
+	objMeta.HTTPMeta["content-type"] = objInfo.ContentType
+	objMeta.HTTPMeta["content-encoding"] = objInfo.ContentEncoding
+
+	for key, val := range objInfo.UserDefined {
+		objMeta.HTTPMeta[key] = val
+	}
+
+	objMeta.Anonymous = anon
+	objMeta.Version = diskCacheObjectMetaVersion
+	return objMeta
+}
+
+// Disk caching
+type diskCache struct {
+	dir string // caching directory (--cache-dir)
+	// Max disk usage limit (--cache-max) in percent. If the disk usage crosses this
+	// limit then we stop caching.
+	maxUsage int
+	// Cache revalidation (in days)
+	expiry int
+	// for storing temporary files
+	tmpDir string
+	// for storing cached objects (renamed from tmp location)
+	dataDir string
+	// creation-time of the cache
+	createTime time.Time
+	// BoltDB for storing atime
+	db *bolt.DB
+
+	// purge() listens on this channel to start the cache-purge process
+	purgeChan chan struct{}
+}
+
+// Returns if the disk usage is low.
+// Disk usage is low if usage is < 80% of maxUsage
+// Ex. for a 100GB disk, if maxUsage is configured as 70% then maxUsage is 70G
+// hence disk usage is low if the disk usage is less than 56G (because 80% of 70G is 56G)
+func (c diskCache) diskUsageLow() bool {
+	minUsage := c.maxUsage * 80 / 100
+
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		errorIf(err, "Error getting disk information on %s", c.dir)
+		return false
+	}
+	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	return int(usedPercent) < minUsage
+}
+
+// Retusn if the disk usage is high.
+// Disk usage is high if disk used is > maxUsage
+func (c diskCache) diskUsageHigh() bool {
```

Comment:
Can above two functions be combined? as single function which basically gives high/low behavior based on a boolean input?

Reply:
no, its more readable with Low/High words than with boolean, also implementation is clearer with separate functions